source/agora/common/BanManager.d

```diff
@@ -180,13 +178,20 @@ public class BanManager
 
     public void onFailedRequest (Address address, uint fail_count_inc = 1) @safe nothrow
     {
-        auto status = this.get(address);
-        if (this.isBanned(address) || status.whitelisted)
+        // fast return if already banned or whitelisted
+        if (this.isBanned(address) || address in whitelisted)
             return;
 
-        status.fail_count += fail_count_inc;
+        try
+        {
+            this.url_failures.require(address).fail_count += fail_count_inc;
+        }
+        catch (Exception e)
+        {
+            log.trace("Exception setting fail_count for address {}", address);
+        }
```

Comment:
If `require` fails, later access will, too. Additionally, every indexing means a lookup, which is an amortized constant, but still. `require` returns a `ref`, take the address of it.

Reply:
Return early when fails and now store instance as you suggested.